### ISSUE TITLE
adv with imu

### ISSUE THREAD

**Issue by MarianaRibas** [dankelley/oce#887]
I’m interested in ADV data treatment and I see you have a function for that. I would like to know how do you deal with data correction for movement, when the ADV have an IMU incorporate. My ultimate goal is to compute TKE to correct the CO2 gas transfer velocity. I'm using two Nortek ADV mounted in the same drifting structure, one with IMU and one without. I'm more than happy to provide more info and data and help on what I can :)

**Comment by dankelley** [comment]
Thanks. Can you send a private email to me and to Clark Richards, containing (or giving a *private* ftp/dropbox link to) a dataset and also any documentation you have on the IMU data stream?  (I am not clear on whether the IMU data are contained within the ADV data file, etc.) We will keep your data private.

**Comment by richardsc** [comment]
Hi Mariana. Feel free to send data that we can have a look at, but the short answer is that we don't do anything in oce to correct for sensor motion. In principle it shouldn't be too difficult, as long as it's not too hard to read the IMU data.

**Comment by MarianaRibas** [comment]
Hi Dan and Clark,

IMU data is indeed contained on the ADV data file, but only the one that have the IMU. The second one without IMU(slave) we need somehow to use it from the first one (master). Both ADV were synchronize before deployment. I do have the vectors from the head ADV to the structure. 

 

In the shared folder I upload example of two dataset from the same day, the one starting with “I” is the one with IMU. The ones of use are .vec, other are info. I have also attached some pdf with documentation, I think one of the most useful is Kilcher adv-turb:  https://www.dropbox.com/sh/gqdmil7ftykzcq0/AADleq71mYpRfOS-Vq4Mhzy4a?dl=0 

 

I thank you in advance and I’m happy to provide more info, clarification, data or help in any way.

 

Cheers, 

 

Mr2

**Comment by dankelley** [comment]
Hi Mariana, please note that I deleted some things from your comment. It is best to use the github web interface, and not the email interface, because otherwise a lot of extra material gets published to the web (often, phone numbers from email signature lines, etc). In this case, I also removed your note on the dropbox link, on the assumption that the data are private.

I cannot see in the pdf files how the MMU data are being incorporated into the `.vec` file (or some other file). What we would need would be a document that explained this, in order to help. Or maybe the MMU info is in one of the other files ... we could help more quickly if we knew such things.

**Comment by dankelley** [comment]
A note to Clark - I got a more recent nortek document that mentions the inertial microstrain unit (IMU) and put it into shared space `oce-working-notes/nortek/system-integrator-manual_Dec2014_jan.pdf`; also, I've downloaded the python code (https://github.com/lkilcher/dolfyn) to see how they do it. It seems that the magic flag is `0x71`, from a quick grep in the python code, but I've not yet seen where this is mentioned in `oce-working-notes/nortek/system-integrator-manual_Dec2014_jan.pdf`).

**Comment by MarianaRibas** [comment]
I had a look into .hdr files and the only different seems in .dat files, the one with imu has:

19   Ensemble counter

20   Timer                            (s)

21   DeltaAngle X                     (deg)

22   DeltaAngle Y                     (deg)

23   DeltaAngle Z                     (deg)

24   DeltaVelocity X                  (m/s)

25   DeltaVelocity Y                  (m/s)

26   DeltaVelocity Z                  (m/s)

27   Orientation Matrix 1,1

28   Orientation Matrix 1,2

29   Orientation Matrix 1,3

30   Orientation Matrix 2,1

31   Orientation Matrix 2,2

32   Orientation Matrix 2,3

33   Orientation Matrix 3,1

34   Orientation Matrix 3,2

35   Orientation Matrix 3,3

36   Heading                          (deg)

37   Pitch                            (deg)

38   Roll                             (deg)

Not sure if that what you were asking.

**Comment by dankelley** [comment]
The table on page 31 of `system-integrator-manual_Dec2014_jan.pdf` suggests that the IMU data are flagged by the code `0xa5` followed by `0x71`. This does not look to be very hard to incorporate into oce, actually. Actually *doing* something with the data is another matter, and probably will end up involving some time-series work as well as geometrical conversion.

@MarianaRibas - do you have the ability to build R packages from source? You'll need to be able to do that to test any solution that gets coded into `oce`, because *official* releases of `oce` are only once or twice per year.

**Comment by dankelley** [comment]
I've made a new branch called `nortek_imu` and I'll probably see about some trial coding tonight or tomorrow morning.

**Comment by dankelley** [comment]
@MarianaRibas do you know that the file named `I5080902.VEC` actually contains IMU data? 

@richardsc I pushed to the branch named `nortek_imu` some test code. And in the oce-issues site [887a.R](https://github.com/dankelley/oce-issues/blob/master/887/887a.R) I made a test file that will only work for the three of us (since it relies on a shared dropbox folder for the data). When I run this code, it gets only 2 "hits" for INU data, whether I try to read just a few samples or a whole lot of them. So clearly I'm not getting things right. But I think it's also possible the .VEC file does not contain INU data. (I guess it would make sense to try the python code as a test.) I'll be going offline now for a couple of hours.  FYI, I've been working in `R/adv.nortek.R` and `src/bitwise.c`. I note that the flags for this data chunk are a bit odd, e.g. they do checksums differently than for other chunks.  Another oddness is that there are **three** types of INU chunk, and we'd have to look at byte offset 5 to find which it is. That subdivision can be handled by looking at a third keying byte, although I've not bothered with that because I'm not getting the hits I would have expected. I did some twiddling manually with e.g. `which(buf==0xa5)` to find the first byte of the chunk and `which(buf==0x71)` to find the second. Well, I don't see many sequences where the two bytes occur in pairs, so I really do not think this `I5080902.VEC` file contains INU data, hence the question for @MarianaRibas  above.

**Comment by richardsc** [comment]
As a test I downloaded the two testfiles from the [Dolfyn examples](https://github.com/lkilcher/dolfyn/tree/master/example_data). Presumably the file labeled `_imu*` has IMU data in it so that could be a good place to start.

One thing I notice though is that there is a difference in reading the "non imu" file from reading the "imu" one:
```r
> library(oce)
> d <- read.oce('vector_data01.VEC', to=1000)
> plot(d) #works
> d <- read.oce('vector_data_imu01.VEC', to=1000)
Warning message:
NAs introduced by coercion 
> plot(d)
Error in oce.plot.ts(x@data$time, y, ylab = beamName(x, which[w]), drawTimeRange = drawTimeRange,  : 
  x must be ordered
```

**Comment by dankelley** [comment]
I'm going to look at all of this tomorrow morning. A bit tired now. I was in RStudio for something else, and tried to run the code and my gawd, I dislike that Rstudio interface -- everything I type seems to be wrong.

**Comment by dankelley** [comment]
In the example, two times are equal. I'll look tomorrow into how that can happen. I've improved the error message. Now, if they are equal, it spits as below. If they are unordered, it prints the values. This way it will be less mysterious.
```
Error in oce.plot.ts(x@data$time, y, ylab = beamName(x, which[w]), drawTimeRange = drawTimeRange,  : 
  x must be distinct but x[42]=x[43]=1339502404
```

**Comment by dankelley** [comment]
Actually, and this is an indication that I'm tired, I don't see why `oce.plot.ts()` should even demand that the data be ordered. Sheesh ... pretty demanding, that function! I'll fix when my eyes are focussing better.

**Comment by MarianaRibas** [comment]
Hi,

Sorry I was away from office yesterday, buried from meeting after meeting. I try to answer all question in one email:

 

-           I don’t know about “ability to build R packages from source” work but if you give me some hints or tutorial I can try. 

-          do you know that the file named I5080902.VEC actually contains IMU data? Yes, this is what I understood but I could be wrong. This is the only input phyton code ask for. Another file where IMU data is (I think) is .dat

Let me know if I can be of further help,

 

Mr2

**Comment by dankelley** [comment]
1. A procedure for building `oce` from source is described at the 'source-code development version' section of http://dankelley.github.io/oce/installation.html

2. I also have been very busy and have not had the time to look into reading these INU-enabled files.  And, in any case, all we would then have would be the ability to read the INU data, but not to process them in any scientifically-meaningful way.

3. I think the nuances of processing the data will require a lot more work. These data are not very common (yet) and so that work would be likely to be in the publishable category, i.e. something that would need clear and careful treatments, along with peer-review, etc.  As I understand things at the moment, the literature on this sort of thing is still fairly thin. Are you working with someone who is an expert in these things?

4. If you want results in a fairly short timescale, you should probably use the python package dolfyn. That saves you from waiting for oce to be able to read the INU data stream, and it may give you access to the advanced processing methods (of joining the two data streams) that I think you will need.  I return to my question at the end of item 3 ... you will probably need to rely on another member of your team who is more familiar with this sort of data. I assume that someone wrote a grant to install the INU device, and to pay technicians to help in deployment, etc., and that that person would have broad experience in working with INU data. From our communications, I gather that your own research field is a bit separate from this.

**Comment by dankelley** [comment]
I started work on this in `src/bitwise.c`, a function named `locate_vector_imu_sequences`. There is some test code in the comments above that, which show how I am testing this (without building oce).  All this does at the moment is to recognize the byte sequences, yielding a nominal false-positive rate of about 1 part in 10^12. I want to test the checksum also, but there seems to be no information in `system-integrator-manual_Dec2014_jan.pdf` on how those checksums are calculated.  (I did not check the docs too carefully, however.)

**Comment by dankelley** [comment]
The `nortek_imu` branch now recognizes imu data chunks and extracts the timestamps, which are in an unit described in documents published by the imu manufacturer but not described in the nortek docs. A test of this is in the [887b.R](https://github.com/dankelley/oce-issues/blob/master/887/887a.R). Note that these dolfyn test files might have the serial number in a different format than other files (the summary comes out odd) but this is not my main concern now.

I plan to make the code extract all the information from one type of imu file. There are three types, but I want to get clear that I'm doing things correctly on one type before going to the others. I'm using the dolfyn test file.

**Comment by dankelley** [comment]
oce now has code to read all the IMU data (only for type 0xcc though). This is as of commit 55474f5df742e7b4aaeeedb366456386b9846bb3 in the `vector_imu` branch. Below is what I get for [887c.R](https://github.com/dankelley/oce-issues/blob/master/887/887c.R), but note that I have not checked to see if any of it is correct.
* Time ranges from 2012-06-12 12:00:02 to 2012-06-12 12:14:01 with 27043 samples and mean step 0.03102581 s
* Statistics of data
```
                                  Min.          Mean         Max.
     v [m/s]               -3.57900000  -0.293328230   0.82700000
     a                     82.00000000 103.482823651 140.00000000
     q                     32.00000000  95.877121621  99.00000000
     pressure [dbar]      198.57000000 472.882506638 623.27800000
     recordsBurst           0.00000000   0.000000000   0.00000000
     voltageSlow           10.80000000  11.017878427  11.30000000
     headingSlow [°]        0.00000000   0.000000000   0.00000000
     pitchSlow [°]          0.00000000   0.000000000   0.00000000
     rollSlow [°]           0.00000000   0.000000000   0.00000000
     temperatureSlow [°C]  10.60000000  11.481954708  12.14000000
     IMUaccelx [m/s²]      -2.31906128  -0.996993137   1.15597951
     IMUaccely [m/s²]      -0.26996461   0.009609649   0.20072642
     IMUaccelz [m/s²]      -0.35694209  -0.010285940   0.46752042
     IMUangrtx [°/s]       -0.99437290  -0.004668213   0.94414389
     IMUangrty [°/s]       -0.24736676  -0.000318323   0.25401562
     IMUangrtz [°/s]       -0.08619732   0.002448380   0.11107717
     IMUmagrtx [gauss]      0.52970427   0.550985042   0.56074685
     IMUmagrty [gauss]     -0.32696378  -0.041999171   0.09039580
     IMUmagrtz [gauss]     -0.20537320  -0.078558822   0.10251021
     IMUM11                -0.10820094  -0.007852737   0.06194739
     IMUM12                -0.10279949  -0.035552452   0.10157529
     IMUM13                 0.99205673   0.997555529   0.99999964
     IMUM21                -0.99998927   0.109154085   0.99983591
     IMUM22                -0.33654514   0.542327183   0.99999803
     IMUM23                -0.06221964  -0.016802081   0.01490101
     IMUM31                -0.99997813  -0.540657313   0.33880097
     IMUM32                -0.99999624   0.107889914   0.99778277
     IMUM33                -0.12557825   0.008905482   0.10439564
     IMUtime [s]            0.06113600 419.212140000 838.36313600
```

**Comment by dankelley** [comment]
I just made [887d.R](https://github.com/dankelley/oce-issues/blob/master/887/887d.R) and it graphs x, y and z accelerations as below. I've no real idea of what should make sense but the numbers are clearly not random. Note that times are in seconds past ... well, I dunno. I don't know how these sync up with the normal nortek vector data. For now, the `adv` object just stores them as-is.

![887d](https://cloud.githubusercontent.com/assets/99469/13680653/fb70fdfc-e6cf-11e5-8848-e6277536bc0b.png)

**Comment by MarianaRibas** [comment]
Is this still only working with example files, not with my file? I assume last notes are just informative, right?

**Comment by dankelley** [comment]
Yes, these notes are just how I keep track of my progress. My next step will be to try your sample file but that won't likely be until Sunday because of busy days until then. And, remember, all that's being done so far is to try to read the imu data, not use it in any particular analysis.

**Comment by dankelley** [comment]
As a followup on last point: Mariana's files do not get detected by the present code as being IMU files. (see [886a.R](https://github.com/dankelley/oce-issues/blob/master/887/887a.R) for a test.) I do not know why. I coded the detection method according to the Nortek documentation, and things seem okay for the dolfyn test files.

**Comment by dankelley** [comment]
I instrumented the code to see what the 5th byte in the sequence is, for Mariana's file. It is 0xc3. This is not in the list of three possibilities described in the Nortek manual (`system-integrator-manual_Dec2014_jan.pdf`), so `read.adv.nortek()` is ignoring these sequences. I don't see how it can do anything else, really, since the data format within a sequence depends on this 5th byte. We need an updated manual. @MarianaRibas do you have a more recent manual? If so, please email it to me and I should be able to get the code working.

**Comment by dankelley** [comment]
FYI this 0xc3 case is handled by dolfyn <https://github.com/lkilcher/dolfyn/blob/master/dolfyn/io/nortek.py#L616> but I prefer to have some documentation to refer to; copying undocumented code is not part of the oce way.

**Comment by MarianaRibas** [comment]
I have asked Nortek for updated manual. Thanks!

**Comment by MarianaRibas** [comment]
Dan, here the answer from Nortek. I let you when I have the updated manual:

 

“You are correct, the Integration Manual is not properly updated and this will be done not too far into next week. Let me know if this is a matter of urgency though.

Have you seen this FAQ <http://cs.nortek.no/scripts/customer.fcgi?_sf=0&custSessionKey=&customerLang=en&noCookies=true&action=viewKbEntry&id=107> , covering the IMU subject? I hope you are aware of the importance of using the newer firmware and software, as a serious bug was found in the earliest versions.”

**Comment by richardsc** [comment]
Hi @MarianaRibas 

Just based on the link that Nortek provided above, I think you might have a firmware issue:

> Before continuing; be aware that the description below applies for Vectors using firmware/software versions from V3.42/V1.39 and onward. If you use older versions, an update is recommended.

and it looks like the firmware version for your Nortek is:
```> d <- read.oce('15080902.vec')
> str(d)
Formal class 'adv' [package "oce"] with 3 slots
  ..@ metadata     :List of 44
  .. ..$ filename                : chr "/Users/richardsc/Downloads/15080902.vec"
  .. ..$ units                   :List of 6
  .. ..$ manufacturer            : chr "nortek"
  .. ..$ instrumentType          : chr "vector"
  .. ..$ numberOfSamples         : int 32500
  .. ..$ numberOfBeams           : int 3
  .. ..$ samplingRate            : num 32
  .. ..$ serialNumber            : chr "VEC11089"
  .. ..$ softwareVersion         : num 1.39
  .. ..$ firmwareVersion         : num 3.36
```

Not sure how this would impact your measurements, but you might want to consider contacting Nortek for more details about what the "serious bug" is.

**Comment by dankelley** [comment]
I've coded into `read.adv.nortek` the ability to read the three data formats in the `system-integrator-manual_Dec2014_jan.pdf` document. I only have a test file for the type I call `B`, so the code for types `C` and `D` are provisional. Mariana's file is in the type I call `A`, and if nortek supplies documentation for that type, I'll code it in.  PS. this will just be code to read, not to process.

**Comment by MarianaRibas** [comment]
This is Nortek answer:

“You should definitely update your software and firmware version. The bug was related to a timing sync issue between the IMU and the Vector. By doing ASCII conversion of data will reveal the problem, and it is unfortunately not possible to recover the data afterwards. We did send out a warning to customers with IMU sensors, recommending an update, but I understand that this email did not reach you and that is unfortunate”

**Comment by dankelley** [comment]
Mariana, I think you should write back to nortek for more clarity on this.  It is not clear from what they said whether you can use the ascii files. If the IMU times are wrong, maybe you can just use the non-IMU times, assuming the instruments sampled at the same time. They say "sync", though, and that might mean that the IMU data are simply unrelated to the non-IMU data.  You haven't said anything about deployment characteristics or scientific approaches, so I don't know whether this means you need to discard your data or not. This is not an oce issue, of course.  Clark and I are both scientists and we could maybe help in trying to do some science with the data but that would be more in the line of coauthors than as oce developers.

PS. Over the weekend, I made a guess as to what is contained in your .VEC file and some of the data seem to make some sense (e.g. every time slice of the 3D rotation matrix has matrix determinant equal to 1) but the IMU times were indeed strange, and unlike the dolfyn sample file. It could be that I was reading them wrong, though. I would still like to get the nortek system-integrator document update, so we could close this issue and have oce being able to read the sort of files you will get in future.

**Comment by MarianaRibas** [comment]
Dan, I wrote back to Nortek with your comments, I updated you when I received the answer. Also I have asked again for the document update J

I’m more than happy to offer you a coauthorship if you could help me clarify this physics mess J

**Comment by MarianaRibas** [comment]
Dan, I added to dropbox shared folder the updated System Integrator Manual.

Also I copy the answer for Nortek: 

“At the first look of it it looks like the instrument did use the newest firmware, a version that do not have the bug mentioned earlier. So that is good news! The files seems to convert properly also, so there is reason to be optimistic about the data you have collected.

I know the support for IMU systems is not as good as we would like it to be, but a lot of the knowledge and processing around these sensors needs to be left with our users. This is something we have been open with from the day the IMU option was released.”

**Comment by dankelley** [comment]
Can you state the file name in the directory?

**Comment by MarianaRibas** [comment]
system-integrator-manual_Mar2016.pdf

**Comment by dankelley** [comment]
I don't see it there:
```
$ ls -c1 ~/Dropbox/OCE-ADV-IMU-Mr2
15080902.dat
15080902.hdr
15080902.pck
15080902.sen
15080902.ssl
15080902.vec
ComprehensiveManual.pdf
Correction of velocities from two Nortek ADV using a single IMU.pdf
I5080902.VEC
I5080902.dat
I5080902.hdr
I5080902.pck
I5080902.sen
I5080902.ssl
I5080902.vhd
I5080902_ENU.dat
Kilcher et al adv_turb_manual DRAFT.pdf
Kilcher_Thomson_Colby_Coherence_METS2014.pdf
Orientation of vectors.pdf
Thomson_etal_GMREC-METS_2013_turbulencemooring.pdf
eric_siegel Vector IMU.pdf
nortek-vector-and-microstrain-support-memo-july.pdf
```

**Comment by MarianaRibas** [comment]
https://www.dropbox.com/sh/gqdmil7ftykzcq0/AADleq71mYpRfOS-Vq4Mhzy4a?dl=0   <https://github.com/notifications/beacon/AQ4fTxthKyvd3JFbIS4WBswE_VPTK8F4ks5ptqnkgaJpZM4Hquju.gif>

**Comment by dankelley** [comment]
Thanks. I don't know how it happened, but my dropbox was not updating for this directory. I have it now, thanks.

**Comment by dankelley** [comment]
Now that I see the docs (page 32 of `system-integrator-manual_Mar2016.pdf`) I can verify that the code I wrote on the weekend was correct. The times it is finding are odd, though, as illustrated below. In `oce` it is assumed that the time increment of the IMU is 16 microseconds. I cannot see anywhere in `system-integrator-manual_Mar2016.pdf` (certainly not on page 32) an indication of the actual meaning of the time stamp. I don't see how to reconcile the timestamps between this and the 
```r
> library(oce)
Loading required package: gsw
> d1 <- read.adv.nortek("~/Dropbox/OCE-ADV-IMU-Mr2/15080902.vec")
> vectorShow(d1[['time']])
[1] "d1[[\"time\"]]: 2015-08-09 06:00:02.015, 2015-08-09 06:00:02.046, ..., 2015-08-09 15:01:32.265, 2015-08-09 15:01:32.296 (length 32500)\n"
> 
> dI <- read.adv.nortek("~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC")
Warning message:
In read.adv.nortek("~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC") :
  IMU type 'A' is not handled properly yet; requested Nortek documentation, March 10, 2016
> vectorShow(dI[['time']])
[1] "dI[[\"time\"]]: 2015-08-09 06:00:12.015, 2015-08-09 06:00:12.046, ..., 2015-08-09 15:52:22.265, 2015-08-09 15:52:22.296 (length 35540)\n"
> vectorShow(dI[['IMUtime']])
[1] "dI[[\"IMUtime\"]]: 0.058384, 0.089392, ..., 0.306384, 0.337392 (length 35540)\n"
```

**Comment by dankelley** [comment]
I plotted the IMU timestamp versus the Vector timestamp:
```r
look <- 1:100
oce.plot.ts(dI[['time']][look], dI[['IMUtime']][look], type='o')
```
and it is as below. Again, I don't know what to make of this.
![screen shot 2016-03-15 at 10 47 04 am](https://cloud.githubusercontent.com/assets/99469/13779775/774d3c1e-ea9b-11e5-8fca-fffe34a0424c.png)

**Comment by dankelley** [comment]
Another odd thing -- the older Nortek docs had several types of IMU data (signalled by the value of `AHRS ID`) but the new document states there is only one type. Maybe the old types are now defunct. I didn't purchase the device from Nortek so I don't think I really have much excuse to pose questions to them on such things.

**Comment by MarianaRibas** [comment]
We did pay a lot of money for these instrument so I can pose all the questions we need J

I keep you update with Nortek’s answers, let me know if I can help any further.

**Comment by dankelley** [comment]
At this moment, I am most interested to know answers to the following specific questions.

1. Will `.vec` files employ `0xcc`, `0xd2` and `0xd3` for AHRS, as documented on pages 31 and 32 of `system-integrator-manual_Dec2014_jan.pdf`, or *only* `0xc3` as documentd on page 32 of `system-integrator-manual_Mar2016.pdf`.

2. What is the meaning of the IMU timestamp? Specifically, what does a value of 0 indicate, and what does a value of 1 indicate?  (I had inferred from microstrain documents that the value of 1 would mean 16 microseconds, but this contradicts what I see in your file. I have not seen anywhere a definition of 0.) You could point them to the diagram in comment https://github.com/dankelley/oce/issues/887#issuecomment-196825367 for an example, for your data.

**Comment by dankelley** [comment]
I would also ask Nortek:

3. How is 'DeltaAngle' defined?

4. How is 'DeltaVelocity' defined?

**Comment by MarianaRibas** [comment]
Dan, Nortek answers: 

·         Will .vec files employ 0xcc, 0xd2 and 0xd3 for AHRS, as documented on pages 31 and 32 of system-integrator-manual_Dec2014_jan.pdf, or only 0xc3 as documented on page 32 of system-integrator-manual_Mar2016.pdf.

The only option is to select dVel, dAng and Orientation matrix as output from the IMU (which is defined as 0xc3). The other options were only available in the previous versions.



·         What is the meaning of the IMU timestamp? Specifically, what does a value of 0 indicate, and what does a value of 1 indicate? (I had inferred from microstrain documents that the value of 1 would mean 16 microseconds, but this contradicts what I see in your file. I have not seen anywhere a definition of 0.) The last graph I send is an example, for my data.

All we do with the timestamp here is to divide it by 62500. I do not understand your question, I think. Timestamp is a 4 byte value (from offset 66-70 in the binary data). Please elaborate.

**Comment by MarianaRibas** [comment]
1.      How is 'DeltaAngle' defined?

2.      How is 'DeltaVelocity' defined?

Check out Microstrains Data communication protocol for a good description of these parameters (Page 65): http://files.microstrain.com/3DM-GX3-15,-25%20MIP%20Data%20Communications%20Protocol.pdf

**Comment by dankelley** [comment]
Thanks for the information, Mariana. I imagine I'll have a look at this over the weekend -- busy until then. Note that the nortek docs (p32 of the `system-integrator-manual_Mar2016.pdf`) state angles in degrees, but the new file you kindly provided states them in radians. This is the sort of thing that is discouragingly common.

If I had bought the machine, I think I would ask Nortek to update to `system-integrator-manual_Mar2016.pdf`, stating the unit correctly. I would also ask them to state the unit for the timestamp. And, finally, they should state the *zero* of that timestamp. The point of all of this is that there are customers who want to work with the original data, and they need clear and correct documentation.

**Comment by dankelley** [comment]
@MarianaRibas I have instrumented the `oce` code to print out the relevant bytes for your file.  Below I am putting the results, as part A. Then, in part B, I verify that the bytes in the file are as the R code is finding them. Note that in part A, both big-endian and little-endian reading are tested.

The question for Nortek (if you can supply them with the file) is: (a) do they agree on the bytes that the `oce` code is discovering, and (b) do they then agree with the two discovered timestamps, of 0.058384 and 0.089392 seconds? If they agree, then we have this extra question, going back to the diagram that I think you sent them, of how the timestamp goes up and then down... is it getting reset somehow?

**Part A** R test
```
> library(oce)
> d<-read.oce("~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC")
IMU entry number: 1
    starting byte (B, say) in file: 1789
    byte[B+0  =1789]: 0xa5 ... check: should be 0xa5
    byte[B+1  =1790]: 0x71 ... check: should be 0x71
    byte[B+5  =1794]: 0xc3 ... check: should be 0xc3
    byte[B+66 =1855]: 0x41
    byte[B+67 =1856]: 0x0e
    byte[B+68 =1857]: 0x00
    byte[B+69 =1858]: 0x00
timestamp (in seconds) if little endian: 0.058384
timestamp (in seconds) if big endian:    17462.984704
IMU entry number: 2
    starting byte (B, say) in file: 1885
    byte[B+0  =1885]: 0xa5 ... check: should be 0xa5
    byte[B+1  =1886]: 0x71 ... check: should be 0x71
    byte[B+5  =1890]: 0xc3 ... check: should be 0xc3
    byte[B+66 =1951]: 0xd3
    byte[B+67 =1952]: 0x15
    byte[B+68 =1953]: 0x00
    byte[B+69 =1954]: 0x00
timestamp (in seconds) if little endian: 0.089392
timestamp (in seconds) if big endian:    -12057.575424
```

**Part B** check the bytes in unix
```
$ od -x -j 1788 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003374      00a5                                                        
0003375
oce[nortek_imu*]$ od -x -j 1789 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003375      0071                                                        
0003376
oce[nortek_imu*]$ od -x -j 1854 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003476      0041                                                        
0003477
oce[nortek_imu*]$ od -x -j 1855 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003477      000e                                                        
0003500
oce[nortek_imu*]$ od -x -j 1856 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003500      0000                                                        
0003501
oce[nortek_imu*]$ od -x -j 1857 -N 1 ~/Dropbox/OCE-ADV-IMU-Mr2/I5080902.VEC
0003501      0000                                                        
0003502

```

**Comment by dankelley** [comment]
@MarianaRibas -- if you contact Nortek, please cut/paste from the previous comment FROM THE WEBPAGE, not the email, because I had an error in the original one and the correction will not have been emailed to you. Thanks.

**Comment by dankelley** [comment]
@MarianaRibas as an update, the code
```r
col.names <- c("burst", "ensemble",
               "v1", "v2", "v3",
               "a1", "a2", "a3",
               "snr1", "snr2", "snr3",
               "cor1", "cor2", "cor3", 
               "p",
               "ai1","ai2",
               "checksum",
               "ensembleCounter", 
               "timer",
               "da1", "da2", "da3", "dv1", "dv2", "dv3",
               "m11", "m12", "m13",
               "m21", "m22", "m23",
               "m31", "m32", "m33",
               "heading", "pitch", "roll")

d <- read.table("I5080902.dat", header=FALSE, col.names=col.names)
d[1:2,]
plot(d$burst[1:100], d$timer[1:100], type='o')
```
verifies the `timer` values, and gives a graph as below. Therefore I think I am reading the bytes correctly. The questions for Nortek is: what, exactly, how is the `timer` defined? Why, in this case, does it repeat in this way, whereas in the `dolfyn` sample file, it increments monotonically?

<img width="1031" alt="screen shot 2016-03-17 at 3 33 32 pm" src="https://cloud.githubusercontent.com/assets/99469/13856788/a417ee22-ec55-11e5-8988-b717eb65c63b.png">

**Comment by dankelley** [comment]
Now that I am gaining confidence that `read.adv.nortek` is reading some of the data correctly, I decided to make some tests against the Nortek-created text files. This is [887g.R](https://github.com/dankelley/oce-issues/blob/master/887/887g.R) in the git-issue repo. There are the usual difficulties of the text file having only a few digits, but there are also some things that just seem wildly wrong.  The two on this thread can run the code easily -- and it's not for any other readers anyway -- and they should be aware that I plan to plug away at this on the weekend. 

My present feeling is that the nortek-supplied text files are just wrong, because the R code is using the correct byte offsets (increasing by 4) and data are okay before and after the odd things.

**Comment by MarianaRibas** [comment]
Nortek response: 

“My apologies, the manual is now updated, based on the information in the Microstrain manual. 

Be also aware that the microstrain coordinate system differs from the Vectors, but I am sure you have read about this already. For your reference, attached is a .m file describing the calculations we do in our software, and this file can be used to convert from XYZ to ENU based on the transformation matrix found in the .dat file. This assumes a Vector with a fixed stem, or to be more precise, the XYZ coordinates we have for a fixed stem Vector, relative to the IMU.”

 

I have asked Nortek your other questions. 

 

About your time questions, it just occurs to me, that in order to synchronize the two vectors we select in the non-IMU vector burst interval so the slave with start with the first burst interval. See a screen cut of deployment planning. Can that explain this up and down? Did you found the same behavior with the two vectors?

 




clear all
load IMU101.dat
d = IMU101;
indMat = 27:35;
M = d(:,indMat);
L = length(M);

V_xyz = d(:,3:5);
V_enu = zeros(size(V_xyz));

A = [0, 0, -1; ...
     0, 1, 0; ...
     1, 0, 0];

B = [0, 1, 0; ...
     1, 0, 0; ...
     0, 0, -1];
 
for i=1:L,
    Rimu = reshape(d(i,indMat),3,3)';
    Omat = A*Rimu*B;
    V_enu(i,:) = transpose(Omat)*V_xyz(i,:)';
       
    Mcalc = Omat;
    roll(i) = atan2(Omat(2,3),Omat(3,3))*180/pi;
    pitch(i) = -asin(Omat(1,3))*180/pi;
    heading(i) = 90.0 - atan2(Omat(1,2),Omat(1,1))*180/pi;
    if heading(i) < 0
        heading(i)=heading(i)+360;    
    end
end

figure(3)
plot(V_enu)
grid
legend('E','N','U')

load IMU101_ENU.dat
hold on
plot(IMU101_ENU(:,3:5))

**Comment by dankelley** [comment]
Thanks. I knew about the rotation code. The prime question right now is about `deltaVelocityY` and `deltaVelocityZ`, which are very different in the binary `.VEC` file than in the text `.dat` file. Following the link to https://github.com/dankelley/oce-issues/blob/master/887/887g.R shows this. I don't see how my code can be reading other things well, and just these two things incorrectly. For example (and I think you should quote this to nortek): The `.dat` file has the following for the start of `deltaVelocityZ` (in my notation)
```
-0.29789 -0.29734 -0.29677 -0.29800 -0.29689 -0.29947
```
However, when I read things in `oce`, I get
```
-0.03037681 -0.03032058 -0.03026246 -0.03038788 -0.03027449 -0.03053777
```
Clearly, these are **wildly** different. There are similar, but smaller, problems for the `y` component of delta-velocity. I don't see how my code can be reading things incorrectly, given that it reads correctly the items before and after Y and Z.  (By 'correctly', I mean that the test code checks every single value for equality, with a tolerance matching the degraded resolution provided in the `.dat` file.)

Nortek people can read my code for `deltaVelocityY` at
https://github.com/dankelley/oce/blob/nortek_imu/R/adv.nortek.R#L168
and the line after is for `Z`, which is even worse.

I don't see how my code can get these particular 8 bytes wrong. Since I cannot see their code, I cannot offer an informed opinion on how the results of my code differ from the `.dat` file. Perhaps Nortek support people can answer that. Whether R is used or something else is used, you will need to be certain of the data. At this stage, I think we have a large problem.

As for the timestamp, I have no idea. I guess it doesn't matter, because we have agreement between R and the `.dat` file on the values, and those values are not useful. The only solution may be to use the clock in the vector itself.  Pending information from nortek on the "zero" of the microstrain clock, that may be the best approach anyway.

**Comment by dankelley** [comment]
@MarianaRibas by "My apologies, the manual is now updated, based on the information in the Microstrain manual." do they mean they sent you the update? If so, can you put it into the dropbox? The reason I ask is that `oce` docs generally make precise statements about source materials.

But, as I say, the very highest priority is figuring out about these delta-velocities, as in my previous comment.

**Comment by dankelley** [comment]
@MarianaRibas re "About your time questions, it just occurs to me, that in order to synchronize the two vectors we select in the non-IMU vector burst interval so the slave with start with the first burst interval. See a screen cut of deployment planning. Can that explain this up and down? Did you found the same behavior with the two vectors?" was that written to me, or by nortek to you? I assume they were to you -- I was not there when the instruments got set up for the experiment.

**Comment by MarianaRibas** [comment]
https://www.dropbox.com/sh/gqdmil7ftykzcq0/AADleq71mYpRfOS-Vq4Mhzy4a?dl=0 system-integrator-manual_Mar2016.pdf (same name as before but updated)

**Comment by MarianaRibas** [comment]
Dan, the difference between two line is always 9.806, is “g” as state in p.32  of system-integrator file. Is this make sense with other deltaVelocity?

**Comment by MarianaRibas** [comment]
Here Nortek answer:

“a) We use the little endian architecture. From my point of view (I am not a programmer) it looks correct. The byte numbers and the checks you run is consistent with what we expect.
b) Yes, this looks correct. The reset happens for every new burst.

"This is a timer value which measures the time since system power-up or timer set command (in our case: new burst). The timer interval on the 3DM-GX3-15 and 3DM-GX3-25 IMU/AHRS is 16 microseconds (μs). To convert the timer value to time in seconds, divide by 62,500. The system clock has an accuracy of +/- 0.01%. The timer value rolls over from its maximum value to 0 approximately every 68719 seconds (~1145 minutes or ~19 hours). When the timestamp is included in the data message format with other IMU/AHRS data quantities, the timestamp represents the time at which the data quantities were sampled. See the Data Synchronicity section of this manual for more details." - this is taken from the IMU manual that I sent yesterday, but is in the newer version of our System integrator included

Does dolfyn example contain continuous measurements or burst data?”

 

 

Here my comment: Our sensors contain burst data for correct synchronization between then.

**Comment by dankelley** [comment]
The new nortek manual is a great improvement. Thanks for encouraging them to rewrite this part.

It is still confusing on the units of `Delta Velocity Rate x`, which seem to be `(g*seconds)` which really makes no sense. I see that they copied the notation from the microstrain document, and at least the latter is a bit clearer.

I am still at a loss to understand why what I read from the `.VEC` file for `Delta Velocity X` (and Y and Z) do not agree with the values in the `.dat` file. Maybe they have flipped them from the microstrain coordinate system to a nortek system. At this stage it is guesswork, and I don't see much benefit in guessing.

**I suggest you ask Nortek this question (verbatim)**: "Please state the formula that relates the delta-velocity values in the textual .dat file to the corresponding values in the binary .VEC file."

I personally think Nortek should put the answer in the manual. And they should do the same for *any* variable that is not identical in the two files. This will save them a lot of emails from customers (and, potentially, a loss of customers to other companies).

**Comment by MarianaRibas** [comment]
Here Nortek answer: 

“When doing the binary to ASCII data conversion in the software, the coordinate transformation from XYZ_imu and XYZ_vector is aligned, as mentioned earlier. Also, the multiplication of g on velocity is performed. The System Integrator Manual applies to the raw data file, whild the format of the ASCII converted .dat file is presented in the .hdr file.

Check out this forum posting describing how to do the transformation: http://www.nortek-as.com/en/knowledge-center/forum/velocimeters/443904138?b_start=0#110572749

I will not put this information in the manual at this point, but will put it on the list of suggested future improvements. The idea with the System Integrators Manual is that it shall cover the need for customers that want to communicate with the instruments using their own controller/own software. If one is using our deployment software instead, the Sys int manual is not the right place to search for information.”

**Comment by dankelley** [comment]
I am opening up a new issue on this single item. Everything else seems to be relatively OK. Note to @MarianaRibas -- it is still advised that you use the text files, because until Nortek provides a clear formula in their documentation, whatever oce is doing is close to a guess.

**Comment by dankelley** [comment]
@MarianaRibas -- I think this issue should be closed. As formally stated, it has been addressed. In oce we try to have focussed issues. As of now, there is one main issue, and that is now addressed in #893. I cannot promise much progress on that issue until Nortek says something specific in their manuals. Dealing with incorrect and changing manuals simply exposes oce users to more confusion than they deserve. I cannot for the life of me understand why a profit-seeking company would choose to release incorrect manuals (e.g. the radian-versus-degree error), but that is their choice. The choice in oce is to try to serve the users well, and for that reason development effort tends to centre on documented data types.

Mariana, I think you need to be dealing with the text files, and not trying to read the `.VEC` file. That way, you have more power to pressure Nortek to answer questions.

None of this, of course, gets you close to the science issue of what to do with the data, but that is not an oce issue, but rather one that might relate to assembling expertise and funding.

**Comment by dankelley** [comment]
I'm closing this because it (as posed in the title) seems to have been addressed, and there has been no action on the issue for nearly three weeks.
